src/edits.rs

```diff
@@ -228,14 +237,19 @@ where
                     },
                     minus_section,
                 ));
-                annotated_plus_line.push((
-                    if coalesce_space_with_previous {
-                        plus_op_prev
-                    } else {
-                        noop_insertion
-                    },
-                    plus_section(n, &mut y_offset),
-                ));
+                let op = if coalesce_space_with_previous {
+                    plus_op_prev
+                } else {
+                    noop_insertion
+                };
+                let plus_section = plus_section(n, &mut y_offset);
+                let line = plus_section.trim_end_matches(&[' ', '\t', '\n']);
+                if !line.is_empty() && line != plus_section.trim_end_matches('\n') {
+                    annotated_plus_line.push((op, line));
+                    annotated_plus_line.push((plus_op_prev, &plus_section[line.len()..]));
+                } else {
+                    annotated_plus_line.push((op, plus_section));
+                }
```

Comment:
@wescande Thanks very much for wading into what must be the most complex/obscure part of the codebase! I haven't fully paged in how this all works. Can you help me: it looks like there's some quite similar logic going on here and on line 95 above; why is that and is it possible to do it just once rather than twice?

Also, WDYT about writing it something like the following, to make it self-documenting and easier to understand for future readers:

```rust
if let Some(non_whitespace) = get_contents_before_trailing_whitespace(plus_section)
{
    annotated_plus_line.push((op, non_whitespace));
    annotated_plus_line.push((plus_op_prev, &plus_section[non_whitespace.len()..]));
} else {
    annotated_plus_line.push((op, plus_section));
}

fn get_contents_before_trailing_whitespace(line: &str) -> Option<&str> {
    let line = line.trim_end_matches(&[' ', '\t', '\n']);
    if !line.is_empty() && line != line.trim_end_matches('\n') {
        Some(line)
    } else {
        None
    }
}
```

Reply:
Agreed this is not an easy part

Yes we can definitively do that only once, I did it when trying to make a more complex split with the `" \t"`.
I uploaded the change accordingly. It is not really smaller but it will easier for re-reader :+1: